tiledb/sm/array_schema/dynamic_array.h

```diff
@@ -0,0 +1,361 @@
+ *
+ * This class does not fill the gap exactly, however. Filling the gap would
+ * require that the array elements be default-constructed in place, as `array`
+ * does. This class does not require that its content type be
+ * default-constructible, since that would make it ineligible for use with a
+ * C.41-compliant class without a default constructor. This class is more
+ * general, allowing non-default constructible content types.
+ *
+ * This class, therefore, fills the gap as a special case by allowing
+ * default-construction of its contents even as it does not mandate
+ * default-construction. Responsibility for initialization with its user. This
+ * class is a combination of (1) uninitialized storage and (2) policy-based
+ * initialization at construction time.
+ *
+ * In addition, this class does not require its allocator to be
+ * default-constructible. This means that users of the class might have
+ * difficulty using swap to implement move-construction, since there may be no
+ * natural way to make a nonce object to swap with.
+ *
+ * @tparam T The type of the element used in the array
+ * @tpara Alloc Allocator for storage. Defaults to the library allocator.
+ */
+template <class T, class Alloc = tiledb::common::allocator<T>>
+class DynamicArray {
+  /**
+   * Allocator instance for type T
+   */
+  Alloc a_;
+
+  /**
+   * The number of allocated objects.
+   *
+   * The allocated size in bytes is `size_*sizeof(T)`.
+   */
+  size_t size_;
+
+  /**
+   * `data_` is an allocated pointer to hold a sequence of `T` whose length is
+   * equal to `size`. It's allocated with the class template argument `Alloc`.
+   *
+   * There's no invariant that data_ not be null. There are two cases,
+   * however, when it might be null. The first is during move construction,
+   * when it's assigned null to achieve transfer semantics. The second is when
+   * a container is constructed with zero size (presumably as a placeholder).
+   */
+  T* data_;
+
+ public:
+  /**
+   * Size type [container.requirements.general]
+   */
+  using size_type = size_t;
+
+  /**
+   * The null initializer does no initialization, but does compile as a tagged
+   * template argument for the constructor.
+   *
+   * @param item Pointer to uninitialized memory in which to construct an object
+   * @param index Index in the array of this item
+   */
+  class NullInitializer {
+   public:
+    inline static void initialize(
+        [[maybe_unused]] T* item, [[maybe_unused]] size_t index){
+        // Arguments ignored; named only to document policy concept
+    };
+  };
+
+  /**
+   * Default initializer policy.
+   *
+   * @param item Address at which to construct an object.
+   */
+  class DefaultInitializer {
+   public:
+    inline static void initialize(T* item, size_t) {
+      new (item) T();
+    };
+  };
+
+  /**
+   * Constructor with no initialization of contained elements.
+   *
+   * The allocator is not called if the size is not positive.
```

Comment:
It should throw an exception rather than default-constructing an object. I avoided throwing only because of the state of exception handling in the code elsewhere.

Reply:
Created SC-15364 to track.